### PR TITLE
Feature/implement tiovx isp methods

### PR DESCRIPTION
**Pull request by jafet-chaves**

### CODE REVIEW

File: gst-libs/gst/tiovx/gsttiovxsimo.c [hunk 1]
```diff
@@ -72,12 +72,49 @@
 #include "gsttiovxmeta.h"
 #include "gsttiovxpad.h"
 #include "gsttiovxutils.h"
+#include "gsttiovxbufferutils.h"
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxsimo.c:75]: Alphabetical order

File: gst-libs/gst/tiovx/gsttiovxutils.h [hunk 2]
```diff
@@ -169,13 +169,12 @@ vx_enum gst_tiovx_get_exemplar_type (vx_reference * exemplar);
  * @exemplar: vx_reference describing the buffer
  * @caps: GstCaps describing the buffer
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.h:170]: Remove caps from documentation

File: gst-libs/gst/tiovx/gsttiovxsiso.c [hunk 3]
```diff
@@ -578,7 +578,7 @@ gst_tiovx_siso_propose_allocation (GstBaseTransform * trans,
     goto exit;
   }
   /* We use input vx_reference to propose a pool upstream */
-  size = gst_tiovx_get_size_from_exemplar (priv->input, priv->in_caps);
+  size = gst_tiovx_get_size_from_exemplar (*priv->input);
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxsiso.c:581]: Does MISO require this change too?

File: gst-libs/gst/tiovx/gsttiovxsimo.h [hunk 4]
```diff
@@ -143,6 +146,29 @@ struct _GstTIOVXSimoClass
  */
 guint	gst_tiovx_simo_get_num_pads      (GstTIOVXSimo *self);
 
+
+#define GST_TYPE_TIOVX_SIMO_QUEUEABLE gst_tiovx_simo_queueable_get_type ()
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxsimo.h:150]: Why is this hardcoded to SIMO? This object should have its own header and its own definition, because it can later be integrated to SISO or MISO. Please separate this from SIMO because it is not a SIMO-only object

File: gst-libs/gst/tiovx/gsttiovxsimo.c [hunk 5]
```diff
@@ -72,12 +72,49 @@
...
 
 #define DEFAULT_BATCH_SIZE (1)
 
 GST_DEBUG_CATEGORY_STATIC (gst_tiovx_simo_debug_category);
 #define GST_CAT_DEFAULT gst_tiovx_simo_debug_category
 
+struct _GstTIOVXSimoQueueable
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxsimo.c:82]: All these methods and logic of this new object shouldn't be part of SIMO, create its own definition

File: gst-libs/gst/tiovx/gsttiovxsimo.c [hunk 6]
```diff
@@ -1377,12 +1431,12 @@ gst_tiovx_simo_process_graph (GstTIOVXSimo * self)
   g_return_val_if_fail (VX_SUCCESS ==
       vxGetStatus ((vx_reference) priv->graph), VX_FAILURE);
 
-  gst_tiovx_pad_get_params (priv->sinkpad, &exemplar, &graph_param_id,
-      &node_param_id);
 
   /* Enqueueing parameters */
   GST_LOG_OBJECT (self, "Enqueueing parameters");
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxsimo.c:1436]: For debugging purposes, in this LOG we can print the graph_param_id and the node_param_id

File: ext/tiovx/gsttiovxisp.c [hunk 7]
```diff
@@ -66,22 +66,60 @@
...
 
+#define DEFAULT_NUM_CHANNELS 1
+#define MAX_SUPPORTED_OUTPUTS 1
+#define DEFAULT_TIOVX_SENSOR_ID "SENSOR_SONY_IMX390_UB953_D3"
+
+
+#define GST_TYPE_TIOVX_ISP_TARGET (gst_tiovx_isp_target_get_type())
+#define DEFAULT_TIOVX_ISP_TARGET TIVX_TARGET_VPAC_VISS1_ID
+
+static const int input_param_id = 3;
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:84]: Let's follow the convention defined. @jafet-chaves have already made a lot of changes to change everything to defines, don't break this again. Move this to defines

File: ext/tiovx/gsttiovxisp.c [hunk 8]
```diff
@@ -240,6 +307,13 @@ static void
 gst_tiovx_isp_init (GstTIOVXISP * self)
 {
   self->dcc_config_file = NULL;
+  /* TODO: this should be a property */
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:310]: You already implemented this property. Remove this TODO comment

File: ext/tiovx/gsttiovxisp.c [hunk 9]
```diff
@@ -312,48 +413,390 @@ gst_tiovx_isp_init_module (GstTIOVXSimo * simo,
...
+  g_return_val_if_fail (context, FALSE);
+  g_return_val_if_fail (sink_pad, FALSE);
+  g_return_val_if_fail (src_pads, FALSE);
+  g_return_val_if_fail (sink_caps, FALSE);
+  g_return_val_if_fail (src_caps_list, FALSE);
+
+  self = GST_TIOVX_ISP (simo);
+
+  tiovx_querry_sensor (&self->sensor_obj);
+  tiovx_init_sensor (&self->sensor_obj, self->sensor_id);
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:433]: You don't seem to be calling the tiovx_deinit_sensor anywhere. You have to call it if:
1. This function fails after this line
2. At the deinit_module function

File: ext/tiovx/gsttiovxisp.c [hunk 10]
```diff
@@ -312,48 +413,390 @@ gst_tiovx_isp_init_module (GstTIOVXSimo * simo,
...
+  self->viss_obj.input.bufq_depth = DEFAULT_NUM_CHANNELS;
+  self->viss_obj.input.params.width = GST_VIDEO_INFO_WIDTH (&in_info);
+  self->viss_obj.input.params.height = GST_VIDEO_INFO_HEIGHT (&in_info);
+  /* TODO: this information should not be hardcoded and instead be obtained from
+   * the query of the sensor
+   */
+  self->viss_obj.input.params.num_exposures = 1;
+  self->viss_obj.input.params.line_interleaved = vx_false_e;
+  self->viss_obj.input.params.format[0].pixel_container = TIVX_RAW_IMAGE_16_BIT;
+  self->viss_obj.input.params.format[0].msb = 11;
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:459]: Where is this msb field obtained from? Is this custom or based on the dcc file? if so, we need to expose this as a property

File: ext/tiovx/gsttiovxisp.c [hunk 11]
```diff
@@ -312,48 +413,390 @@ gst_tiovx_isp_init_module (GstTIOVXSimo * simo,
...
+
+  self->viss_obj.input.bufq_depth = DEFAULT_NUM_CHANNELS;
+  self->viss_obj.input.params.width = GST_VIDEO_INFO_WIDTH (&in_info);
+  self->viss_obj.input.params.height = GST_VIDEO_INFO_HEIGHT (&in_info);
+  /* TODO: this information should not be hardcoded and instead be obtained from
+   * the query of the sensor
+   */
+  self->viss_obj.input.params.num_exposures = 1;
+  self->viss_obj.input.params.line_interleaved = vx_false_e;
+  self->viss_obj.input.params.format[0].pixel_container = TIVX_RAW_IMAGE_16_BIT;
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:458]: This looks hardcoded isn't it?

File: ext/tiovx/gsttiovxisp.c [hunk 12]
```diff
@@ -312,48 +413,390 @@ gst_tiovx_isp_init_module (GstTIOVXSimo * simo,
...
+  self->viss_obj.input.params.format[0].msb = 11;
+  self->viss_obj.input.params.meta_height_before = 0;
+  self->viss_obj.input.params.meta_height_after = 0;
+
+  self->viss_obj.ae_awb_result_bufq_depth = DEFAULT_NUM_CHANNELS;
+
+  GST_INFO_OBJECT (self,
+      "Input parameters: \n  Width: %d \n  Height: %d \n  Pool size: %d",
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:466]: Print all the configuration, don't be shy! It is helpful for debugging purposes

File: ext/tiovx/gsttiovxisp.c [hunk 13]
```diff
@@ -312,48 +413,390 @@ gst_tiovx_isp_init_module (GstTIOVXSimo * simo,
...
+  }
+
+  self->viss_obj.output2.bufq_depth = DEFAULT_NUM_CHANNELS;
+  self->viss_obj.output2.color_format =
+      gst_format_to_vx_format (out_info.finfo->format);
+  self->viss_obj.output2.width = GST_VIDEO_INFO_WIDTH (&out_info);
+  self->viss_obj.output2.height = GST_VIDEO_INFO_HEIGHT (&out_info);
+
+  self->viss_obj.h3a_stats_bufq_depth = DEFAULT_NUM_CHANNELS;
+
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:500]: Print output configuration as you are doing it for the input

File: ext/tiovx/gsttiovxisp.c (hunk 11 above)
Comment (rrcarlosrodriguez) [reply ext/tiovx/gsttiovxisp.c:458]: We should get it from input caps somehow (?)

File: ext/tiovx/gsttiovxisp.c [hunk 14]
```diff
@@ -312,48 +413,390 @@ gst_tiovx_isp_init_module (GstTIOVXSimo * simo,
...
 {
-  return FALSE;
+  GstTIOVXISP *self = NULL;
+  GList *l = NULL;
+  GstTIOVXSimoQueueable *queueable_object = NULL;
+  gint graph_parameter_index = 0;
+
+  g_return_val_if_fail (simo, FALSE);
+  g_return_val_if_fail (sink_pad, FALSE);
+  g_return_val_if_fail (src_pads, FALSE);
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:564]: Pointer guard for the queueable_objects list

File: ext/tiovx/gsttiovxisp.c [hunk 15]
```diff
@@ -312,48 +413,390 @@ gst_tiovx_isp_init_module (GstTIOVXSimo * simo,
...
+    GstTIOVXPad *src_pad = (GstTIOVXPad *) l->data;
+
+    /* Set output parameters */
+    gst_tiovx_pad_set_params (src_pad,
+        (vx_reference) self->viss_obj.output2.image_handle[0],
+        graph_parameter_index, output2_param_id);
+    graph_parameter_index++;
+  }
+
+  /* ae_awb results & h3a stats aren't input or outputs, these are added as queueable_objects */
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:587]: Nice approach :)

Comment (rrcarlosrodriguez) [reply ext/tiovx/gsttiovxisp.c:587]: But again, this object shouldn't be hardcoded to SIMO, it doesn't have to do with SIMO at all

File: ext/tiovx/gsttiovxisp.c [hunk 16]
```diff
@@ -366,5 +809,151 @@ gst_tiovx_isp_compare_caps (GstTIOVXSimo * simo, GstCaps * caps1,
...
+out:
+  if (!ret && *memory) {
+    gst_memory_unref (*memory);
+  }
+
+  return ret;
+}
+
+static gboolean
+gst_tiovx_isp_allocate_user_data_objects (GstTIOVXISP * self)
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:898]: Put a comment to tell the user what is this function's goal

File: ext/tiovx/gsttiovxisp.c [hunk 17]
```diff
@@ -366,5 +809,151 @@ gst_tiovx_isp_compare_caps (GstTIOVXSimo * simo, GstCaps * caps1,
...
+static gboolean
+gst_tiovx_isp_allocate_single_user_data_object (GstTIOVXISP * self,
+    GstMemory ** memory, vx_user_data_object user_data)
+{
+  vx_size data_size = 0;
+  vx_status status = VX_FAILURE;
+  GstTIOVXMemoryData *ti_memory = NULL;
+  gboolean ret = FALSE;
+
+  g_return_val_if_fail (memory, FALSE);
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:849]: Pointer guard for self

Comment (rrcarlosrodriguez) [reply ext/tiovx/gsttiovxisp.c:849]: We should check user_data somehow too

File: ext/tiovx/gsttiovxisp.c [hunk 18]
```diff
@@ -366,5 +809,151 @@ gst_tiovx_isp_compare_caps (GstTIOVXSimo * simo, GstCaps * caps1,
...
+  if (!ret) {
+    GST_ERROR_OBJECT (self, "Unable to allocate data for H3A stats user data");
+  }
+
+out:
+  return ret;
+}
+
+static gboolean
+update2Aresults (vx_user_data_object ae_awb_result)
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:926]: There is a missing logic here (not needed for this release). We need a way to, for each buffer processed, call this function, in order to use the AE library provided by TI and complete the loop, getting the h3a stats and setting the ae params back to the node. Just to have in mind that that loop is incomplete

File: ext/tiovx/gsttiovxisp.c (hunk 7 above)
Comment (emmadrigal) [reply ext/tiovx/gsttiovxisp.c:84]: The standard is to move this to `static const int` as seen here

File: gst-libs/gst/tiovx/gsttiovxsiso.c (hunk 3 above)
Comment (emmadrigal) [reply gst-libs/gst/tiovx/gsttiovxsiso.c:581]: MISO doesn't use this functions

File: gst-libs/gst/tiovx/gsttiovxutils.c [hunk 19]
```diff
@@ -343,34 +368,49 @@ gst_tiovx_empty_exemplar (vx_reference ref)
 
 /* Gets size from exemplar and caps */
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxutils.c:369]: remove caps

File: gst-libs/gst/tiovx/gsttiovxsimo.h [hunk 20]
```diff
@@ -143,6 +146,6 @@ struct _GstTIOVXSimoClass
  */
 guint	gst_tiovx_simo_get_num_pads      (GstTIOVXSimo *self);
 
-G_END_DECLS
+
```
Comment (rrcarlosrodriguez) [inline gst-libs/gst/tiovx/gsttiovxsimo.h:149]: Why did you remove this?

File: ext/tiovx/gsttiovxisp.c [hunk 21]
```diff
@@ -366,5 +944,168 @@ gst_tiovx_isp_compare_caps (GstTIOVXSimo * simo, GstCaps * caps1,
...
+{
+  uint8_t *data_buf;
+  vx_map_id ae_awb_result_map_id;
+
+  vxMapUserDataObject (ae_awb_result,
+      0,
+      sizeof (tivx_ae_awb_params_t),
+      &ae_awb_result_map_id,
+      (void **) &data_buf, VX_WRITE_ONLY, VX_MEMORY_TYPE_HOST, 0);
+  /* Currently just setting it to zero, but will be computed based on previous VISS h3a stats and updated */
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:1088]: Is this properly indented?

File: ext/tiovx/gsttiovxisp.c [hunk 22]
```diff
@@ -366,5 +944,168 @@ gst_tiovx_isp_compare_caps (GstTIOVXSimo * simo, GstCaps * caps1,
...
+  if (!ret) {
+    GST_ERROR_OBJECT (self, "Unable to allocate data for H3A stats user data");
+  }
+
+out:
+  return ret;
+}
+
+static gboolean
+update_2a_results (vx_user_data_object ae_awb_result)
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:1078]: Documentation for this one. Also a pointer guard for ae_awb_result

File: ext/tiovx/gsttiovxisp.c [hunk 23]
```diff
@@ -205,6 +287,58 @@ gst_tiovx_isp_class_init (GstTIOVXISPClass * klass)
...
+  g_object_class_install_property (gobject_class, PROP_LINE_INTERLEAVED,
+      g_param_spec_boolean ("lines-interleaved", "Interleaved lines",
+          "Flag to indicate if lines are interleaved",
+          default_lines_interleaved,
+          G_PARAM_READWRITE | GST_PARAM_CONTROLLABLE | G_PARAM_STATIC_STRINGS |
+          GST_PARAM_MUTABLE_READY));
+
+  g_object_class_install_property (gobject_class, PROP_FORMAT_MSB,
+      g_param_spec_int ("format-msb", "Format MSB",
+          "Flag inditicating which is the most significant bit that still has data",
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:321]: Typo: indicating

File: ext/tiovx/gsttiovxisp.c [hunk 24]
```diff
@@ -300,6 +482,12 @@ gst_tiovx_isp_get_property (GObject * object, guint prop_id,
     case PROP_DCC_CONFIG_FILE:
       g_value_set_string (value, self->dcc_config_file);
       break;
+    case PROP_SENSOR_ID:
+      g_value_set_string (value, self->sensor_id);
+      break;
+    case PROP_TARGET:
+      g_value_set_enum (value, self->target_id);
+      break;
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:490]: There are missing properties

File: ext/tiovx/gsttiovxisp.c [hunk 25]
```diff
@@ -312,48 +500,438 @@ gst_tiovx_isp_init_module (GstTIOVXSimo * simo,
...
+  if (!gst_video_info_from_caps (&in_info, sink_caps)) {
+    GST_ERROR_OBJECT (self, "Failed to get info from input caps: %"
+        GST_PTR_FORMAT, sink_caps);
+    goto out;
+  }
+
+  self->viss_obj.input.bufq_depth = DEFAULT_NUM_CHANNELS;
+  self->viss_obj.input.params.width = GST_VIDEO_INFO_WIDTH (&in_info);
+  self->viss_obj.input.params.height = GST_VIDEO_INFO_HEIGHT (&in_info);
+  /* TODO: this information should not be hardcoded and instead be obtained from
```
Comment (rrcarlosrodriguez) [inline ext/tiovx/gsttiovxisp.c:542]: Change this  TODO message

**Review by rrcarlosrodriguez — CHANGES_REQUESTED** [review]

File: gst-libs/gst/tiovx/gsttiovxsimo.h (hunk 20 above)
Comment (emmadrigal) [reply gst-libs/gst/tiovx/gsttiovxsimo.h:149]: This was a typo
